test/tools/Modules/HttpListener/HttpListener.psm1

```diff
@@ -0,0 +1,168 @@
+Function Stop-HTTPListener {
+    <#
+    .Synopsis
+        Stop HTTP Listener used for PowerShell tests rather than rely on 3rd party websites
+    .Description
+        Sends `exit` command to HTTP Listener causing it to exit.
+    .Parameter Port
+        Port to use, default is 8080
+    .Example
+        Stop-HTTPListener -Port 8080
+    #>
+    Param (
+        [Parameter()]
+        [Int] $Port = 8080
+    )
+
+    Invoke-WebRequest -Uri "http://localhost:$port/PowerShell?test=exit"
+}
+
+Function Start-HTTPListener {
+    <#
+    .Synopsis
+        Creates a new HTTP Listener to be used for PowerShell tests rather than rely on 3rd party websites
+    .Description
+        Creates a new HTTP Listener supporting several test cases intended to be used only by PowerShell tests.
+        This cmdlet requires running from an elevated administrator prompt to open a port.
+
+        Use Ctrl-C to stop the listener.  You'll need to send another web request to allow the listener to stop since
+        it will be blocked waiting for a request.
+    .Parameter Port
+        Port to listen, default is 8080
+    .Example
+        Start-HTTPListener -Port 8080
+        Invoke-WebRequest -Uri "http://localhost:8080/PowerShell/?test=linkheader&maxlinks=5"
+    #>
+    Param (
+        [Parameter()]
+        [Int] $Port = 8080,
+
+        [Parameter()]
+        [switch] $AsJob
+        )
+
+    Process {
+        $ErrorActionPreference = "Stop"
+
+        [scriptblock]$script = {
+            param ($Port)
+
+            # HttpListener.QueryString is not being populated, need to follow-up with CoreFx, workaround is to parse it ourselves
+            Function ParseQueryString([string]$url)
+            {
+                Write-Verbose "Parsing: $url"
+                $queryItems = @{}
+                foreach ($segment in $url.Split("&"))
+                {
+                    if ($segment -match "\??(?<name>\w*)(=(?<value>.*?))?$")
+                    {
+                        $name = $matches["name"]
+                        $value = $matches["value"]
+                        if ($value -ne $null)
+                        {
+                            $value = [System.Web.HttpUtility]::UrlDecode($value)
```

Comment:
Maybe it is better to run entire $url through UrlDecode first, and then to splitting and parsing?